sqs/subscriber.go

```diff
@@ -134,52 +146,69 @@ func (s *Subscriber) ConsumeMessages(
 	output chan *message.Message,
 	logFields watermill.LogFields,
 ) {
-
 	for _, sqsMsg := range messages {
-		ctx, cancelCtx := context.WithCancel(ctx)
-		defer cancelCtx()
-		msg, err := s.config.Unmarshaler.Unmarshal(&sqsMsg)
-		if err != nil {
-			s.logger.Error("Cannot unmarshal message", err, logFields)
+		processed := s.processMessage(ctx, logFields, sqsMsg, output, queueURL)
+
+		if !processed {
 			return
 		}
-		msg.SetContext(ctx)
-		output <- msg
+	}
+}
 
-		select {
-		case <-msg.Acked():
-			err := s.deleteMessage(ctx, queueURL, sqsMsg.ReceiptHandle)
-			if err != nil {
-				s.logger.Error("Failed to delete message", err, logFields)
-				return
-			}
-		case <-msg.Nacked():
-			// Do not delete message, it will be redelivered
-		case <-s.closing:
-			s.logger.Trace("Closing, message discarded before ack", logFields)
-			return
-		case <-ctx.Done():
-			s.logger.Trace("Closing, ctx cancelled before ack", logFields)
-			return
+func (s *Subscriber) processMessage(ctx context.Context, logFields watermill.LogFields, sqsMsg types.Message, output chan *message.Message, queueURL string) bool {
+	s.logger.Trace("processMessage", logFields)
+
+	ctx, cancelCtx := context.WithCancel(ctx)
+	defer cancelCtx()
+
+	msg, err := s.config.Unmarshaler.Unmarshal(&sqsMsg)
+	if err != nil {
+		s.logger.Error("Cannot unmarshal message", err, logFields)
+		return false
+	}
+	msg.SetContext(ctx)
+	output <- msg
+
+	select {
+	case <-msg.Acked():
+		err := s.deleteMessage(ctx, queueURL, sqsMsg.ReceiptHandle, logFields)
+		if errors.Is(err, context.Canceled) {
+			return false
+		}
+		if err != nil {
+			s.logger.Error("Failed to delete message", err, logFields)
+			return false
 		}
+	case <-msg.Nacked():
+		// Do not delete message, it will be redelivered
+		return false // we don't want to process next messages to preserve order
```

Comment:
we were not doing it earlier, but I guess that it's expected? if we'll continue we can process next messages and we may lose ordering

eventually we can add config here